gnosis/verification.k

```diff
@@ -6,21 +6,35 @@ module VERIFICATION
     imports EDSL
     imports LEMMAS
 
+  // ########################
+  // Memory Usage
+  // ########################
+
+    rule #memoryUsageUpdate(#memoryUsageUpdate(MU, START0, WIDTH0), START1, WIDTH1) => #memoryUsageUpdate(MU, START0, WIDTH0)
+      requires START1 +Int WIDTH1 <Int START0 +Int WIDTH0
+
+    rule #memoryUsageUpdate(#memoryUsageUpdate(MU, START0, WIDTH0), START1, WIDTH1) => #memoryUsageUpdate(MU, START1, WIDTH1)
+      requires START0 +Int WIDTH0 <=Int START1 +Int WIDTH1
+
+    rule 0 <=Int #memoryUsageUpdate(MU, START, WIDTH) => true
+
+    rule pow256 <=Int #memoryUsageUpdate(MU, START, WIDTH) => false  requires START +Int WIDTH <Int 3705346855594118253554271520278013051304639509300498049262642688253220148477921
+    rule #memoryUsageUpdate(MU, START, WIDTH) <Int  pow256 => true   requires START +Int WIDTH <Int 3705346855594118253554271520278013051304639509300498049262642688253220148477921
+
+
   // ########################
   // Gas
   // ########################
 
-    // Temporary solution to avoid reasoning about gas
-    syntax Int ::= "#infinity" [function]
- // -------------------------------------
-    rule #infinity +Int _ => #infinity
-    rule #infinity -Int _ => #infinity
-    rule A <=Int #infinity => true
-      requires #getKLabelString(A) =/=String "#infinity_VERIFICATION"
-    rule #infinity <Int A => false
-      requires #getKLabelString(A) =/=String "#infinity_VERIFICATION"
+    syntax Int ::= #gas ( Int , Int ) [function]
+ // --------------------------------------------
+    rule #gas(NONMEM, MEM) -Int #symCmem(MEM') => #gas(NONMEM, MEM +Int MEM')
+    rule #gas(NONMEM, MEM) +Int NONMEM' => #gas(NONMEM -Int NONMEM', MEM)  requires #getKLabelString(NONMEM') =/=String "#symCmem"
+    rule #gas(NONMEM, MEM) -Int NONMEM' => #gas(NONMEM +Int NONMEM', MEM)  requires #getKLabelString(NONMEM') =/=String "#symCmem"
+
+    rule G <=Int #gas(_,_)  => true  requires #getKLabelString(G) =/=String "#gas"
+    rule #gas(_, _) <Int G  => false requires #getKLabelString(G) =/=String "#gas"
```

Comment:
i don't understand these two rules.

Reply:
Used to solve the side condition for these two rules.
```   rule <k> G:Int ~> #deductGas => #end EVMC_OUT_OF_GAS ... </k> <gas> GAVAIL                  </gas> requires GAVAIL <Int G
    rule <k> G:Int ~> #deductGas => .                    ... </k> <gas> GAVAIL => GAVAIL -Int G </gas> <previousGas> _ => GAVAIL </previousGas> requires GAVAIL >=Int G```
GAVAIL will be #gas(_,_) during the execution.

Reply:
i mean, we need to isolate this rule application, so that it is applied only for the semantic rules you mentioned.  otherwise, these rules are not sound.

you may need to change the semantics as well, in order to isolate these.

Reply:
following the above question, if we change #gas takes three arguments, #gas(ORIGIN, NONMEM, MEM), then, we can say like:
```
G <= Int #gas(#infinity, _, _) => true
```

Reply:
will #infinity ever go out side #gas(#infinity, _, _) expression? why we need getKLabel(G) =/= #infinity?

Reply:
i don't know, i'm just giving an example.

my biggest concern for this is when we call an external function (where we pass the currently available gas) and return from it (where we receive the remaining gas).  then, `G` could be`#gas(...)` or some complex term that includes #gas(...) inside.  then, what should be the correct lemma for them?